### DISCUSSION TITLE
[Error name : ValueError] I got some error during coding Help please

### DISCUSSION THREAD

**Discussion by husfsj** [kernc/backtesting.py#605] [Q&A]
below are my strategy code
it use MACD, EMA simple
I have some problem with code that when i run and got this kind of message
 **backtesting.py", line 713, in new_order raise ValueError(ValueError: Long orders require: SL (38498.388) < LIMIT (38293.637299999995) < TP (38846.265)**

i know whats meaning about error but dont know how to fix it 
just guess it have problem with setting SL/TP
can you guys solve this error?
thanks

this is error message

  File "c:\Users\ssungjae\Dropbox\PC\Desktop\Python_auto\MACD_backtesting.py", line 82, in <module>
    stats = bt.run()
  File "C:\Users\ssungjae\Anaconda3\lib\site-packages\backtesting\backtesting.py", line 1170, in run
    strategy.next()
  File "c:\Users\ssungjae\Dropbox\PC\Desktop\Python_auto\MACD_backtesting.py", line 59, in next
    self.buy(sl = long_sl, tp = long_tp, limit = None)
  File "C:\Users\ssungjae\Anaconda3\lib\site-packages\backtesting\backtesting.py", line 210, in buy
    return self._broker.new_order(size, limit, stop, sl, tp)
  File "C:\Users\ssungjae\Anaconda3\lib\site-packages\backtesting\backtesting.py", line 713, in new_order
    raise ValueError(
ValueError: Long orders require: SL (38377.374) < LIMIT (37722.729699999996) < TP (38724.157499999994)

### 
import ccxt
import time
import pandas as pd
import pprint
import numpy
import talib
from backtesting import Strategy
access = ""        
secret = ""        

binanceX = ccxt.binance(config={
    'apiKey': access, 
    'secret': secret,
    'enableRateLimit': True,
    'options': {
        'defaultType': 'future'
    }
})
# load candle data
def GetOhlcv(binance, Ticker, period):
    btc_ohlcv = binance.fetch_ohlcv(Ticker, period)
    df = pd.DataFrame(btc_ohlcv, columns=['datetime', 'Open', 'High', 'Low', 'Close', 'Volume'])
    df['datetime'] = pd.to_datetime(df['datetime'], unit='ms')
    df.set_index('datetime', inplace=True)
    return df
df_5min = GetOhlcv(binanceX, 'BTC/USDT', '5m')

class MACD_EMA(Strategy):
    ### EMA 간격 
    n50 = 50
    n200 = 200
    df_5min = GetOhlcv(binanceX, 'BTC/USDT', '5m')

    
    # MACD mode: macd=0,  macdSignal=1, macdHist=2
    MACD_mode=2

    def init(self):
      # self.MACDhist_5m = self.I(MACD, df_5min['close'])
        self.MACD = self.I(talib.MACD, self.df_5min['Close'])
        self.MACD = self.MACD[self.MACD_mode]

        self.ema50 = self.I(talib.EMA, self.df_5min['Close'], self.n50)
        self.ema200 = self.I(talib.EMA, self.df_5min['Close'], self.n200)

    def next(self):
        if self.ema200[-2] < self.ema50[-2]: #5m EMA 50이 200보다 클 때 ==> only long만
            if self.MACD[-2] < 0 and self.MACD[-3] < 0 and self.MACD[-4] < 0 and self.MACD[-4] < self.MACD[-3] and self.MACD[-3] < self.MACD[-2] and not self.position.is_long and not self.position.is_short:
                # self.buy(sl = self.long_sl, tp =self.long_tp)
                price = self.df_5min["Open"][-1]
                long_sl = price * 0.996
                long_tp = price * 1.005    

                self.buy(sl = long_sl, tp = long_tp, limit = None)
                # for trade in trades:
                #     if price / trade.entry_price > 1.005:
                #         self.position.close()  
                 

                
        elif self.ema200[-2] > self.ema50[-2]: #5m EMA 50이 200보다 작을 때 ==> only short만
            if self.MACD[-2] < 0 and self.MACD[-3] < 0 and self.MACD[-4] < 0 and self.MACD[-4] < self.MACD[-3] and self.MACD[-3] < self.MACD[-2] and not self.position.is_long and not self.position.is_short:
                price = self.df_5min["Open"][-1]
                short_sl = price * 1.004
                short_tp = price * 0.995
                self.buy(sl = short_sl, tp = short_tp, limit = None)
                # for trade in trades:
                #     if price / trade.entry_price < 0.995:
                #         self.position.close()  

from backtesting import Backtest
bt = Backtest(df_5min, MACD_EMA, margin=0.1, cash=1000, commission=0.003, exclusive_orders=False,)

stats = bt.run()
bt.plot()

**Comment by kernc** [comment]
> ValueError: Long orders require: SL (38377.374) < LIMIT (37722.729699999996) < TP (38724.157499999994)

https://github.com/kernc/backtesting.py/blob/94d20da85e278102a0cc71b27a3a35b815e11648/backtesting/backtesting.py#L711-L715

Limit price (or current market price for `limit=None`) must be greater than set SL (`sl=`) price for long orders.

**Reply by husfsj** [reply]
thank you for your help i solved that problem.
thank you very much!!
